frontend/packages/console-plugin-sdk/src/utils/useTranslatedExtensions.ts

```diff
@@ -0,0 +1,53 @@
+import * as React from 'react';
+import { TFunction } from 'i18next';
+import { Extension, LoadedExtension } from '@console/dynamic-plugin-sdk/src/types';
+import { isTranslatableString, translateExtensionDeep } from './extension-i18n';
+import useTranslationExt from './useTranslationExt';
+
+/**
+ * `translateExtensionDeep` mutates the extension for translations. We need to store a
+ * semi-permanent mapping of the translation keys values.
+ *
+ * Structured as: { [extension.UID]: { [propertyPathToTranslation]: translationKey } }
+ */
+const translationKeyMap: Record<string, Record<string, string>> = {};
+
+const useTranslatedExtensions = <E extends Extension>(
+  extensions: LoadedExtension<E>[],
+): LoadedExtension<E>[] => {
+  const { t } = useTranslationExt();
+  // TODO: Drop ref and start using useEffect when we are not mutating extensions
+  const lastTRef = React.useRef<TFunction>(null);
+
+  if (t !== lastTRef.current) {
+    extensions.forEach((e) => {
+      const UID = e.uid;
+      translateExtensionDeep(
+        e,
+        (value, path): value is string => {
+          let translatableString = value;
+          if (translationKeyMap[UID]?.[path]) {
+            translatableString = translationKeyMap[UID][path];
+          }
+          return isTranslatableString(translatableString);
+        },
+        (value, key, obj, path) => {
+          if (!translationKeyMap[UID]) {
```

Comment:
We could use [Logical OR assignment (||=)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment)

```suggestion
          translationKeyMap[UID] ||= {};
```

Reply:
I do not believe this is supported in TypeScript 3.8 -- which the console still uses.

Reply:
You are right, this is supported by TS 4+